### PR TITLE
Remove clone trait from translation

### PR DESCRIPTION
**Pull request by ardbiesheuvel**

### CODE REVIEW

File: src/idmap.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ use crate::{
 use core::ptr::NonNull;
 
 /// Identity mapping, where every virtual address is either unmapped or mapped to the identical IPA.
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq)]
```
Comment (qwandor) [inline src/idmap.rs:19]: We may as well continue to implement `Copy` and `Clone` here seeing as `IdTranslation` is cheap to copy, even though it's not required by the interface anymore.

**Review by qwandor — COMMENTED** [review]
Some minor comments below, but a bigger concern is that this creates a less pleasant API for users of the library. With this change they have to keep the `Translation` implementation around somewhere where it will not be moved, which could be awkward. I can see two alternatives:

1. Keep the library as it is. The `Translation` implementations we provide are lightweight so cloning them all over the place isn't an issue. If a user of the library wants to implement a heavier-weight translation and not have the overhead of copying it, they can implement `Translation` on a reference to their struct rather than the struct itself. This will have the same lifetime issues as above, but at least it only affects people who actually need to use a reference rather than cloning.
2. Store the `Translation` in `RootTable` rather than `PageTableWithLevel`, and pass a reference to it to methods on `PageTableWithLevel` as needed. This would mean that `PageTableWithLevel` couldn't implement `Debug` as it currently does, but I think it shouldn't be too bad otherwise. I can try prototyping this option.

**Comment by ardbiesheuvel** [comment]
> Some minor comments below, but a bigger concern is that this creates a less pleasant API for users of the library. With this change they have to keep the `Translation` implementation around somewhere where it will not be moved, which could be awkward.

I'm not sure I follow - why would that be awkward?

 I can see two alternatives:
> 
> 1. Keep the library as it is. The `Translation` implementations we provide are lightweight so cloning them all over the place isn't an issue. If a user of the library wants to implement a heavier-weight translation and not have the overhead of copying it, they can implement `Translation` on a reference to their struct rather than the struct itself. This will have the same lifetime issues as above, but at least it only affects people who actually need to use a reference rather than cloning.

I think doing multiple clone()'s in a recursive function such as map_range() is just really bad form.

> 2. Store the `Translation` in `RootTable` rather than `PageTableWithLevel`, and pass a reference to it to methods on `PageTableWithLevel` as needed. This would mean that `PageTableWithLevel` couldn't implement `Debug` as it currently does, but I think it shouldn't be too bad otherwise. I can try prototyping this option.

Not sure I follow why we would no longer be able to implement Debug, but this sounds like a better option. But I'd still prefer to simply not rely on cloning whatsoever.

**Comment by qwandor** [comment]
> > Some minor comments below, but a bigger concern is that this creates a less pleasant API for users of the library. With this change they have to keep the `Translation` implementation around somewhere where it will not be moved, which could be awkward.
> 
> I'm not sure I follow - why would that be awkward?

Suppose you want to make a function to set up your page table and return it, something like:

```rust
fn make_page_table() -> Mapping<MyTranslation> {
  let translation = MyTranslation {...};
  let mapping = Mapping::new(&translation, 123, 3);
  mapping
}
```

This won't work with this change, because mapping contains a reference to `translation` and so can't outlive it. But returning `translation` alongside it doesn't help either, because it's still being moved, and so the reference would no longer be valid. You'd have to instead put it on a `Box` or something on the heap, or in a static variable, neither which is likely to be very convenient.

> I can see two alternatives:
> 
> > 1. Keep the library as it is. The `Translation` implementations we provide are lightweight so cloning them all over the place isn't an issue. If a user of the library wants to implement a heavier-weight translation and not have the overhead of copying it, they can implement `Translation` on a reference to their struct rather than the struct itself. This will have the same lifetime issues as above, but at least it only affects people who actually need to use a reference rather than cloning.
> 
> I think doing multiple clone()'s in a recursive function such as map_range() is just really bad form.
We could make the constraint `Copy` rather than `Clone` to make it clear that the `Translation` implementation really should be cheap to copy. This would work for references.

> > 2. Store the `Translation` in `RootTable` rather than `PageTableWithLevel`, and pass a reference to it to methods on `PageTableWithLevel` as needed. This would mean that `PageTableWithLevel` couldn't implement `Debug` as it currently does, but I think it shouldn't be too bad otherwise. I can try prototyping this option.
> 
> Not sure I follow why we would no longer be able to implement Debug, but this sounds like a better option. But I'd still prefer to simply not rely on cloning whatsoever.

This approach would avoid the cloning. The only downside is slightly more verbose code passing the translation around, but I'm trying it now and it doesn't look that bad.

**Comment by qwandor** [comment]
Take a look at #20, I think that's a better approach.

**Comment by ardbiesheuvel** [comment]
Superseded by #20

**Pull request closed without merging**